### DISCUSSION TITLE
Package manager for Lingua Franca

### DISCUSSION THREAD

**Discussion by lhstrh** [lf-lang/lingua-franca#1466] [Specifications]
**Disclaimer: these notes are still incomplete and only meant as an initial outline/draft**

## Build properties
Some build configuration aspects are currently captured by "target properties" that are listed at the beginning of the `.lf` source file. We should deprecate some of these target properties and let them become "build properties" to be considered by our new package manager that will take over the responsibility of orchestrating the build process.

### List of target properties to retire and handle with build properties
- `build`
- `cmake-include`
- ...

### Root

`.toml`

## Visualization

- :question: should package manager also be front end for `lfd`?

## Formatting

- `format` command

## Testing
 -`test`command

## Running
 - `run` command

### Running federated programs

# Action items
 - [ ] Settle on name for package manager
 - [ ] Adapt protoype to only be a front-end for `lfc`
 - [ ] Describe platform-specific workflows in detail

**Comment by tanneberger** [comment]
## Building Cpp-Target

Relevant File: [Cmake Generator](https://github.com/lf-lang/lingua-franca/blob/master/org.lflang/src/org/lflang/generator/cpp/CppStandaloneCmakeGenerator.kt)

```mermaid
graph TD
    A[bash-script is invoked] --> B(lfc parses arguments)
    B --> C(LFC generates cpp code)
    C --> D(LFC generates cmake files from command line and target properties)
    D --> E(LFC starts cmake)
    E --> F(cmake and make)
```
### A couple of noteworthy things
- lfc generates .lf-cpp-marker files together with the cmake file. cmake later does a recursive search through the directory structure searching for those files and then adds the CMakeLists.txt in the same folder.
- lfc pulls reactor-cpp
- lfc generates static files
- lfc has a toggle between ros2 and standard cpp
- cmake has all the platform specific code (linux vs apple vs win)


### List of things lfc bakes into the cmake
- projectName
- cmakeBuildType
- externalRuntimePath
- all the source files
- custom compiler
